nx/lib/nx/defn/expr.ex

```diff
@@ -137,18 +137,33 @@ defmodule Nx.Defn.Expr do
     {preds, exprs} = Enum.unzip(clauses)
     {preds, context} = to_exprs(preds)
 
-    [last | exprs] =
+    {broadcasted_clauses, vectorized_axes} =
       [last | exprs]
       |> Enum.map(&Composite.flatten_list([&1]))
       |> Enum.zip_with(&broadcast_clause/1)
-      |> case do
+      |> Enum.unzip()
+
+    [last | exprs] =
+      case broadcasted_clauses do
         # Handle the case where branches don't return anything
         [] -> Enum.map([last | exprs], fn _ -> {} end)
         clauses -> unzip_clauses(clauses)
       end
 
     clauses = Enum.zip(preds, exprs)
-    flatten_to_composite(out, context, exprs, &expr(&1, context, :cond, [clauses, last]))
+
+    out = flatten_to_composite(out, context, exprs, &expr(&1, context, :cond, [clauses, last]))
+
+    if vectorized_axes == [] do
+      out
+    else
+      {result, []} =
+        Composite.traverse(out, vectorized_axes, fn expr, [axes | tail] ->
+          {Nx.vectorize(expr, axes), tail}
+        end)
+
+      result
+    end
```

Comment:
We may use this in `while`as well, so I can see we encapsulating this soon. :)

Reply:
yup. we might even be able to absorb it into `flatten_to_composite` somehow